src/actors/objects/mod.rs

```diff
@@ -596,242 +233,235 @@ impl ObjectsActor {
-        let file_ids = future::join_all(
-            sources
-                .iter()
-                .map(|source| {
-                    let type_name = source.type_name();
-                    let hub = Arc::new(Hub::new_from_top(Hub::current()));
-                    self.download_svc
-                        .list_files(source.clone(), filetypes, identifier.clone(), hub)
-                        .boxed_local()
-                        .compat()
-                        .or_else(move |e| {
-                            // This basically only happens for the Sentry source type, when doing
-                            // the search by debug/code id. We do not surface those errors to the
-                            // user (instead we default to an empty search result) and only report
-                            // them internally.
-                            log::error!(
-                                "Failed to fetch file list from {}: {}",
-                                type_name,
-                                LogError(&e)
-                            );
-                            Ok(Vec::new())
-                        })
-                })
-                // collect into intermediate vector because of borrowing problems
-                .collect::<Vec<_>>(),
-        )
-        .map(|file_ids: Vec<Vec<SourceFileId>>| -> Vec<SourceFileId> {
-            file_ids.into_iter().flatten().collect()
-        });
-
-        let meta_cache = self.meta_cache.clone();
-        let threadpool = self.threadpool.clone();
-        let data_cache = self.data_cache.clone();
-        let download_svc = self.download_svc.clone();
-
-        let file_metas = file_ids.and_then(move |file_ids| {
-            future::join_all(file_ids.into_iter().map(move |file_id| {
-                let source_id = file_id.source_id().to_owned();
-                let source_location = file_id.location();
+        let file_ids = self
+            .clone()
+            .list_files(sources, filetypes, &identifier)
+            .await;
+        let file_metas = self.fetch_file_metas(file_ids, &identifier, scope).await;
+
+        select_meta(file_metas, purpose)
+    }
+
+    /// Collect the list of files to download from all the sources.
+    ///
+    /// This concurrently contacts all the sources and asks them for the files we should try
+    /// to download from them matching the required filetypes and object IDs.  Not all
+    /// sources guarantee that the returned files actually exist.
+    ///
+    /// TODO(flub): Once all the callers are async/await this should take `&self` again
+    /// instead of requiring `self`.
+    ///
+    // TODO(flub): this function should inline already fetch the meta (what
+    // `fetch_file_metas()` does now) for each file.  Currently we synchronise all the
+    // futures at the end of the listing for no good reason.
+    async fn list_files(
+        self,
+        sources: Arc<[SourceConfig]>,
```

Comment:
I guess with async/await, this can also be `&[SourceConfig]` then.

Reply:
I'd love to pair on this with you :) I'm also unsure why I failed to make this take `&self`